src/Broadway/EventStore/DBALEventStore.php

```diff
@@ -231,4 +235,80 @@ private function convertStorageValueToIdentifier($id)
 
         return $id;
     }
+
+    public function visitEvents(EventVisitorInterface $eventVisitor, Criteria $criteria = null)
+    {
+        $statement = $this->prepareVisitEventsStatement($criteria ?: Criteria::create());
+        $statement->execute();
+
+        while ($row = $statement->fetch()) {
+            if ($this->useBinary) {
+                $row['uuid'] = $this->convertStorageValueToIdentifier($row['uuid']);
+            }
+
+            $domainMessage = $this->deserializeEvent($row);
+
+            $eventVisitor->doWithEvent($domainMessage);
+        }
+    }
+
+    private function prepareVisitEventsStatement(Criteria $criteria)
+    {
+        list ($where, $bindValues, $bindValueTypes) = $this->prepareVisitEventsStatementWhereAndBindValues($criteria);
+        $query = 'SELECT uuid, playhead, metadata, payload, recorded_on
+            FROM ' . $this->tableName . '
+            ' . $where . '
+            ORDER BY id ASC';
+
+        $statement = $this->connection->executeQuery($query, $bindValues, $bindValueTypes);
+
+        return $statement;
+    }
+
+    private function prepareVisitEventsStatementWhereAndBindValues(Criteria $criteria)
+    {
+        if ($criteria->getAggregateRootTypes()) {
+            throw new CriteriaNotSupportedException(
+                'DBAL implementation cannot support criteria based on aggregate root types.'
+            );
+        }
+
+        $bindValues = array();
+        $bindValueTypes = array();
+
+        $criteriaTypes = array();
+
+        if ($criteria->getAggregateRootIds()) {
+            $criteriaTypes[] = array('uuid IN (:uuids)');
+
+            if ($this->useBinary) {
+                $bindValues['uuids'] = array();
+                foreach ($criteria->getAggregateRootIds() as $id) {
+                    $bindValues['uuids'][] = $this->convertIdentifierToStorageValue($id);
+                }
+                $bindValueTypes['uuids'] = Connection::PARAM_STR_ARRAY;
+            } else {
+                $bindValues['uuids'] = $criteria->getAggregateRootIds();
+                $bindValueTypes['uuids'] = Connection::PARAM_STR_ARRAY;
+            }
+        }
+
+        if ($criteria->getEventTypes()) {
+            $criteriaTypes[] = array('type IN (:types)');
+            $bindValues['types'] = $criteria->getEventTypes();
+            $bindValueTypes['types'] = Connection::PARAM_STR_ARRAY;
+        }
+
+        if (! $criteriaTypes) {
+            return array('', array(), array());
+        }
+
+        $where = 'WHERE '.join(' AND ', array_map(function (array $criteriaType) {
+            return implode(' OR ', $criteriaType);
```

Comment:
In the code above `$criteriaType` will always contain only 1 value. Would it make sense to just make it a `string` instead of an `array<string>` and get rid of this `array_map()`?